### PR TITLE
Reject port names containing whitespace

### PR DESCRIPTION
**Pull request by L4co77**
- Add `std::isspace` check to `IsAllowedPortName()` to reject port names containing whitespace                                          
 - Update `CreatePort` error message to mention the whitespace constraint                                                                                                                 
 - Add tests for whitespace rejection in port names

### CODE REVIEW

**Review by dsobek — COMMENTED** [review]
Makes sense, just going to give this a test in against MoveIt Pro before approving.

For the forward port, could you evaluate if it makes sense to merge the upstream of behaviortree.cpp into our fork? In particular, I realized that some of the validation logic I pointed to you exists in the [upstream master branch](https://github.com/PickNikRobotics/BehaviorTree.CPP/commit/8f86eb99f1e4eab366fe77bbc676cec997599016#diff-cbe32b6ab05752091d74697476c9e6325cb8939667e37f0031dc1fd599692bf9R10), and so there might be some duplicate work there. I created a separate issue for this https://github.com/PickNikRobotics/moveit_pro/issues/17640.

**Comment by L4co77** [comment]
> Makes sense, just going to give this a test in against MoveIt Pro before approving.
> 
> For the forward port, could you evaluate if it makes sense to merge the upstream of behaviortree.cpp into our fork? In particular, I realized that some of the validation logic I pointed to you exists in the [upstream master branch](https://github.com/PickNikRobotics/BehaviorTree.CPP/commit/8f86eb99f1e4eab366fe77bbc676cec997599016#diff-cbe32b6ab05752091d74697476c9e6325cb8939667e37f0031dc1fd599692bf9R10), and so there might be some duplicate work there. I created a separate issue for this [PickNikRobotics/moveit_pro#17640](https://github.com/PickNikRobotics/moveit_pro/issues/17640).

Commit 8f86eb9 adds comprehensive name validation that fully covers our whitespace check and more. Merging upstream into our fork would make this PR's BT.CPP change redundant. Should I take on [#17640](https://github.com/PickNikRobotics/moveit_pro/issues/17640)?

**Comment by dsobek** [comment]
> > Makes sense, just going to give this a test in against MoveIt Pro before approving.
> > For the forward port, could you evaluate if it makes sense to merge the upstream of behaviortree.cpp into our fork? In particular, I realized that some of the validation logic I pointed to you exists in the [upstream master branch](https://github.com/PickNikRobotics/BehaviorTree.CPP/commit/8f86eb99f1e4eab366fe77bbc676cec997599016#diff-cbe32b6ab05752091d74697476c9e6325cb8939667e37f0031dc1fd599692bf9R10), and so there might be some duplicate work there. I created a separate issue for this [PickNikRobotics/moveit_pro#17640](https://github.com/PickNikRobotics/moveit_pro/issues/17640).
> 
> Commit [8f86eb9](https://github.com/PickNikRobotics/BehaviorTree.CPP/commit/8f86eb99f1e4eab366fe77bbc676cec997599016) adds comprehensive name validation that fully covers our whitespace check and more. Merging upstream into our fork would make this PR's BT.CPP change redundant. Should I take on [#17640](https://github.com/PickNikRobotics/moveit_pro/issues/17640)?

If we still want to get this fixed for `8.10` I think we should apply this more minimal fix. Updating to the most recent state of the upstream behaviortree.cpp repo is a bit too risky of a change. 
No rush on taking on #17640, I've loaded it onto the next iteration so that we can take care of it then. For the forward port, it might be nice to bundle this work together, but if it ends up being too much of a hassle, I think pointing `main` to the same behaviortree.cpp commit hash as we have in `8.10` (to include this change) is fine for the time being.

File: src/xml_parsing.cpp [hunk 1]
```diff
@@ -214,6 +214,14 @@ void BT::XMLParser::PImpl::loadSubtreeModel(const XMLElement* xml_root)
           {
             throw RuntimeError("Missing attribute [name] in port (SubTree model)");
           }
+          if(!IsAllowedPortName(name))
```
Comment (dsobek) [inline src/xml_parsing.cpp]: So this is turning out to be a bit strict. This is saying that `_collapsed` is not a valid port name, which is true, but it is a valid attribute.

https://github.com/PickNikRobotics/BehaviorTree.CPP/blob/c852132b2090564e29982e63573453f28c013e05/src/xml_parsing.cpp#L661-L710

is where ports are being parsed and this function is used there just to separate ports from non-ports. 

This is the reason for the ci failure [here](https://github.com/PickNikRobotics/moveit_pro/actions/runs/23291681390/job/67739393163?pr=17516) (I printed the errors). 

I also saw this when testing the moveit pro PR.

Image

Maybe lets just move the std::any_of statement here and call it a day.

Comment (L4co77) [reply src/xml_parsing.cpp]: thanks, fixed

Comment (dsobek) [reply src/xml_parsing.cpp]: Hey so the current diff is not exactly what I had imagined, I was thinking that the `std::isspace` fn call would be on the line I originally left this comment on. 

Leaving that logic in `IsAllowedPortName` doesn't do anything for us as that function is only ever called in cases where the input will never have spaces, i.e. attributes of an xml tag, where attributes are already white-space delimited.

**Review by dsobek — COMMENTED** [review]
See https://github.com/PickNikRobotics/BehaviorTree.CPP/pull/18#discussion_r2997146031

**Comment by L4co77** [comment]
@dsobek could you check? thanks

File: src/xml_parsing.cpp (hunk 1 above)
Comment (dsobek) [reply src/xml_parsing.cpp]: I pushed up a commit for what I described here.

**Comment by dsobek** [comment]
Since I added a commit here, I should no longer be the person approving. Added @shaur-k to review.

**Comment by L4co77** [comment]
@shaur-k could you check this ? thanks

File: src/xml_parsing.cpp [hunk 2]
```diff
@@ -214,6 +218,13 @@ void BT::XMLParser::PImpl::loadSubtreeModel(const XMLElement* xml_root)
           {
             throw RuntimeError("Missing attribute [name] in port (SubTree model)");
           }
+          std::string_view sname(name);
+          if(std::any_of(sname.begin(), sname.end(),
+                         [](unsigned char c) { return std::isspace(c); }))
+          {
+            throw RuntimeError(
+                StrCat("The name of a port must not contain whitespace: '", sname, "'"));
+          }
```
Comment (shaur-k) [inline src/xml_parsing.cpp]: deduplicate this to one helper function that gets called in both `basic_types.h` and `xml_parsing.cpp`

**Review by shaur-k — CHANGES_REQUESTED** [review]

File: src/xml_parsing.cpp (hunk 2 above)
Comment (L4co77) [reply src/xml_parsing.cpp]: fixed

File: src/xml_parsing.cpp [hunk 3]
```diff
@@ -1000,10 +1008,59 @@ void BT::XMLParser::PImpl::recursivelyCreateSubtree(const std::string& tree_ID,
...
+              {
+                new_bb->set(port_name, dbl_val);
+                stored = true;
+              }
+            }
+          }
+          if(!stored)
+          {
+            new_bb->set(port_name, str_value);
+          }
```
Comment (JWhitleyWork) [inline src/xml_parsing.cpp:1063]: There are a lot of `convertFromString` functions in `basic_types.hpp` which duplicate much of the logic that you've done here. I think it would be better to use those.

**Review by JWhitleyWork — CHANGES_REQUESTED** [review]
Just one comment.

File: src/xml_parsing.cpp (hunk 3 above)
Comment (L4co77) [reply src/xml_parsing.cpp:1063]: Thanks! I tried this but` convertFromString<T>` doesn't fit three requirements here:                                                   
1. **Strict validation** -` convertFromString<int64_t>` doesn't check `ptr == end`, so `"1;2;3"` silently returns `1` and breaks vector-literal  port values.
2. **Non-throwing fallback** - the try-int/try-double/fallback-to-string pattern needs `std::optional`-style returns; `convertFromString<T>` throws.                                                                                                                              
3. **Int vs. double filter** - the . / `e` / `E` check has to happen before the parse to pick the target type, so we'd need it anyway.

Happy to extract `tryParseStrictInt64` / `tryParseStrictDouble` helpers into `basic_types.h` in a follow-up if reuse is the goal, but the lenient existing functions would regress vector-literal port values.

Comment (JWhitleyWork) [reply src/xml_parsing.cpp:1063]: OK, thanks for the explanation! Approving.

**Review by JWhitleyWork — APPROVED** [review]
